engine/src/main/scala/cromwell/engine/workflow/lifecycle/ShadowMaterializeWorkflowDescriptorActor.scala

```diff
@@ -0,0 +1,282 @@
+
+  private def buildWorkflowDescriptor(id: WorkflowId,
+                                      sourceFiles: WorkflowSourceFiles,
+                                      conf: Config): ErrorOr[EngineWorkflowDescriptor] = {
+
+    val namespaceValidation = validateNamespace(sourceFiles.wdlSource)
+    val workflowOptionsValidation = validateWorkflowOptions(sourceFiles.workflowOptionsJson)
+    (namespaceValidation |@| workflowOptionsValidation) {
+      (_, _)
+    } flatMap { case (namespace, workflowOptions) =>
+      buildWorkflowDescriptor(id, sourceFiles, namespace, workflowOptions, conf)
+    }
+  }
+
+  private def buildWorkflowDescriptor(id: WorkflowId,
+                                      sourceFiles: WorkflowSourceFiles,
+                                      namespace: NamespaceWithWorkflow,
+                                      workflowOptions: WorkflowOptions,
+                                      conf: Config): ErrorOr[EngineWorkflowDescriptor] = {
+    val rawInputsValidation = validateRawInputs(sourceFiles.inputsJson)
+    val failureModeValidation = validateWorkflowFailureMode(workflowOptions, conf)
+    val backendAssignmentsValidation = validateBackendAssignments(namespace.workflow.calls, workflowOptions)
+    (rawInputsValidation |@| failureModeValidation |@| backendAssignmentsValidation ) {
+      (_, _, _)
+    } flatMap { case (rawInputs, failureMode, backendAssignments) =>
+      buildWorkflowDescriptor(id, namespace, rawInputs, backendAssignments, workflowOptions, failureMode)
+    }
+  }
+
+  // TODO: Probably want to enhance this to include any engine-level filesystem and/or filesystem roots
+  case object NoWorkflowEngineFunctions extends WorkflowEngineFunctions(List.empty, new WorkflowContext("")) {
+    override def glob(path: String, pattern: String): Seq[String] =
+      throw new Exception("Cannot use glob in workflow-level declarations because no filesystem yet")
+  }
+
+  private def buildWorkflowDescriptor(id: WorkflowId,
+                                      namespace: NamespaceWithWorkflow,
+                                      rawInputs: Map[String, JsValue],
+                                      backendAssignments: Map[Call, String],
+                                      workflowOptions: WorkflowOptions,
+                                      failureMode: WorkflowFailureMode): ErrorOr[EngineWorkflowDescriptor] = {
+    for {
+      coercedInputs <- validateCoercedInputs(rawInputs, namespace)
+      declarations <- validateDeclarations(namespace, workflowOptions, coercedInputs, NoWorkflowEngineFunctions)
+      declarationsAndInputs = declarations ++ coercedInputs
+      backendDescriptor = BackendWorkflowDescriptor(id, namespace, declarationsAndInputs, workflowOptions)
+    } yield EngineWorkflowDescriptor(backendDescriptor, declarations, backendAssignments, failureMode)
+  }
+
+  private def validateBackendAssignments(calls: Seq[Call], workflowOptions: WorkflowOptions): ErrorOr[Map[Call, String]] = {
```

Comment:
happy to do either, but IMO if we want to go cats we should make that choice soon